src/Lucene.Net/Search/FieldCacheImpl.cs

```diff
@@ -193,57 +191,31 @@ internal Cache(FieldCache wrapper)
             
             internal FieldCache wrapper;
 
-            internal WeakDictionary<object, IDictionary<Entry, object>> readerCache = new WeakDictionary<object, IDictionary<Entry, object>>();
+            internal ConditionalWeakTable<object, ConcurrentDictionary<Entry, object>> readerCache = new ConditionalWeakTable<object, ConcurrentDictionary<Entry, object>>();
             
-            protected internal abstract System.Object CreateValue(IndexReader reader, Entry key, IState state);
+            protected internal abstract T CreateValue(IndexReader reader, Entry key, IState state);
 
             /* Remove this reader from the cache, if present. */
             public void Purge(IndexReader r)
             {
                 object readerKey = r.FieldCacheKey;
-                lock (readerCache)
-                {
-                    readerCache.Remove(readerKey);
-                }
+                readerCache.Remove(readerKey);
             }
             
-            public virtual System.Object Get(IndexReader reader, Entry key, IState state)
+            public virtual T Get(IndexReader reader, Entry key, IState state)
             {
-                IDictionary<Entry, object> innerCache;
-                System.Object value;
-                System.Object readerKey = reader.FieldCacheKey;
-                lock (readerCache)
-                {
-                    innerCache = readerCache[readerKey];
-                    if (innerCache == null)
-                    {
-                        innerCache = new HashMap<Entry, object>();
-                        readerCache[readerKey] = innerCache;
-                        value = null;
-                    }
-                    else
-                    {
-                        value = innerCache[key];
-                    }
-                    if (value == null)
-                    {
-                        value = new CreationPlaceholder();
-                        innerCache[key] = value;
-                    }
-                }
-                if (value is CreationPlaceholder)
+                var innerCache = readerCache.GetOrCreateValue(reader.FieldCacheKey);
+                var value = innerCache.GetOrAdd(key, new CreationPlaceholder<T>());
```

Comment:
Any reason we cannot use `Lazy<T>` here, instead of this?

Reply:
done